lib/Doctrine/ODM/PHPCR/UnitOfWork.php

```diff
@@ -3968,4 +3974,43 @@ public function invokeGlobalEvent($eventName, EventArgs $event)
             $this->eventManager->dispatchEvent($eventName, $event);
         }
     }
+
+    /**
+     * If the parent node has child restrictions, ensure that the given
+     * class name is within them.
+     *
+     * @param NodeInterface $parentNode
+     * @param string $classFqn
+     */
+    private function validateChildClasses(NodeInterface $parentNode, $classFqn)
+    {
+        $parentClass = $this->documentClassMapper->getClassName($this->dm, $parentNode);
+
+        if (null === $parentClass) {
+            return;
+        }
+
+        $metadata = $this->dm->getClassMetadata($parentClass);
+        $childClasses = $metadata->getChildClasses();
+
+        if ($metadata->isLeaf()) {
+            throw new OutOfBoundsException(sprintf(
+                'Document "%s" has been mapped as a leaf. It cannot have children',
+                $parentClass
+            ));
+        }
+
+        if (array() === $childClasses) {
+            return;
+        }
+
+        if (!in_array($classFqn, $childClasses)) {
```

Comment:
indeed, well caught.. at least for subclass.

not sure if it makes sense for interfaces though -- we have no analogy for that in PHPCR - we have mixins ≈ traits and classes = node types, but interfaces?

maybe they could also be modelled by mixins?